src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp

```diff
@@ -0,0 +1,1534 @@
+
+	    for ( nnz_lno_t ii = 0; ii < left_work; ++ii){
+	      size_type a_col = col_begin + ii;
+	      nnz_lno_t rowB = entriesA[a_col];
+	      scalar_t valA = valuesA[a_col] * mult;
+
+	      rowBegin = row_mapB(rowB);
+	      left_workB = row_mapB(rowB + 1) - rowBegin;
+
+	      for ( nnz_lno_t i = 0; i < left_workB; ++i){
+		const size_type adjind = i + rowBegin;
+		nnz_lno_t b_col_ind = entriesB[adjind];
+		scalar_t b_val = valuesB[adjind] * valA;
+		nnz_lno_t hash = b_col_ind & pow2_hash_func;
+
+		hm2.sequential_insert_into_hash_mergeAdd_TrackHashes(hash, b_col_ind, b_val,
+								     &used_hash_sizes, hm2.max_value_size,
+								     &globally_used_hash_count,
+								     globally_used_hash_indices);
+	      }
+	    }
+
+	    for (nnz_lno_t i = 0; i < globally_used_hash_count; ++i){
+	      nnz_lno_t dirty_hash = globally_used_hash_indices[i];
+	      hm2.hash_begins[dirty_hash] = -1;
+	    }
+	  });
+	memory_space.release_chunk(globally_used_hash_indices);
+      }
+
+
+      // This is the default row-based algorithm (thread-sequential)
+      KOKKOS_INLINE_FUNCTION
+      void operator()(const GPUTag&, const team_member_t & teamMember) const {
+
+	nnz_lno_t team_row_begin = teamMember.league_rank()  * team_work_size;
+	const nnz_lno_t team_row_end = KOKKOSKERNELS_MACRO_MIN(team_row_begin + team_work_size, numrows);
+
+	char *all_shared_memory = (char *) (teamMember.team_shmem().get_shmem(shared_memory_size));
+
+	// Shift it to the thread private part
+	all_shared_memory += thread_memory * teamMember.team_rank();
+
+	// Holds the size of 1st and 2nd level hashes
+	volatile nnz_lno_t *used_hash_sizes = (volatile nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * 2;
+
+	nnz_lno_t *globally_used_hash_count = (nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * 2;
+
+	nnz_lno_t * begins = (nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * thread_shmem_hash_size;
+
+	// Points to the next elements
+	nnz_lno_t * nexts = (nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * thread_shmem_key_size;
+
+	// Holds the keys
+	nnz_lno_t * keys = (nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * thread_shmem_key_size;
+
+	// Remainder of shmem allocation for vals
+	scalar_t * vals = KokkosKernels::Impl::alignPtr<char*, scalar_t>(all_shared_memory);
+
+	// Create the hashmaps
+	KokkosKernels::Experimental::HashmapAccumulator<nnz_lno_t,nnz_lno_t,scalar_t>
+	  hm(thread_shmem_hash_size, thread_shmem_key_size, begins, nexts, keys, vals);
+
+	KokkosKernels::Experimental::HashmapAccumulator<nnz_lno_t,nnz_lno_t,scalar_t>
+	  hm2(pow2_hash_size, pow2_hash_size,
+	      NULL, NULL, NULL, NULL);
+
+	Kokkos::parallel_for(Kokkos::TeamThreadRange(teamMember, team_row_begin, team_row_end), [&] (const nnz_lno_t& row_index) {
+
+	    scalar_t mult = -omega*dinv(row_index,0);
```

Comment:
@srajama1 I agree with @seheracer, each row_index is handled exactly once and this is a warp-broadcast access of dinv, so it's already optimal (copying into shared first also accesses the same global addresses once).